src/preprocessing/passes/ackermann.cpp

```diff
@@ -187,9 +187,117 @@ void collectFunctionsAndLemmas(FunctionToArgsMap& fun_to_args,
 
 /* -------------------------------------------------------------------------- */
 
+/* Update the statistics for each uninterpreted sort */
+void updateUSortsCardinality(USortToBVSizeMap& usort_cardinality, TNode term)
+{
+  TypeNode type = term.getType();
+  if (type.isSort())
+  {
+    if (usort_cardinality.find(type) == usort_cardinality.end())
+    {
+      usort_cardinality.insert(make_pair(type, make_pair(0, 0)));
+    }
+    usort_cardinality[type].first = usort_cardinality[type].first + 1;
+  }
+}
+
+/* Given the lowest capacity requirements for each uninterpreted sorts, assign
+ * unique bit vector size. Get the converting map */
+void collectUSortsToBV(std::unordered_set<unsigned>& used,
+                       USortToBVSizeMap& usort_cardinality,
+                       vector<TNode>& vec,
+                       SubstitutionMap& sorts_to_skolem)
+{
+  NodeManager* nm = NodeManager::currentNM();
+
+  for (TNode term : vec)
+  {
+    TypeNode type = term.getType();
+    if (type.isSort())
+    {
+      unsigned size = usort_cardinality[type].second;
+      if (size == 0)
+      {
+        size = log2(usort_cardinality[type].first) + 1;
+        while (used.find(size) != used.end())
+        {
+          ++size;
+        }
```

Comment:
I am not quite sure if it is really necessary to find unique sizes for all the different sorts. If the input is well-typed, we should never be mixing different sort. Maybe I am missing something?

cc @yoni206

Reply:
If your input two (uninterpreted) constants `s1` and `s2` of (corresponding uninterpreted) sorts `S1` and `S2`, then `s1` and `s2` are implicitly distinct.
Shouldn't `S1` and `S2` be mapped to different BV sorts? Otherwise, you might end up constructing a model in which `s1` and `s2` are equal, as this is only an implicit constraint, not an explicit one. Or is it I that is missing something?

Reply:
I haven't thought very deeply about model generation but we would anyway want to map the bit-vectors back to representatives of the corresponding uninterpreted sorts (note our current models doesn't actually include the representatives for uninterpreted sorts, see #2851). To reconstruct the representatives, you just have to look at the bit-vector values of all terms of a given uninterpreted sort and assign a unique representative to each bit-vector value that you see. `s1` and `s2` cannot be equal in that scenario because you are assigning different representatives based on the sort of the original (pre-ackermannization) sort and a term has a unique sort.

Reply:
How can you be sure that `s1` and `s2` will be assigned different representatives? Say `s1` and `s2` are the only variables in the formula. If we map both `S1` and `S2` to `BV[1]`, then `s1` and `s2` will be assigned to variables of the same BV sort. Why are they distinct?

Reply:
Using non-unique size may not change the satisfiability. But for underlying theory, it could be confusing if we replace two different sorts with a same new sort. First, I think the model generation is possible to give s1 and s2 same value. It maybe safe for satisfiability but could be misleading. And Second, consider for further development, converting S1 and S2 to same sort may mislead the optimization algorithm. I mean current method may be safe, but we don't know in the future if there is a smart mechanism that deduce a rule s1=s2 which should not be true. Carefully design may could avoid mistakes, but still, we will cause the system be vulnerable.

Reply:
It could also cause a problem that, the number of sorts and the cardinality for each sort will be ambiguous.

Reply:
Alright, so I think that we all agree that it doesn't change the satisfiability. If the satisfiablity is preserved, then any optimizations that we do on the transformed problem should be safe to do, otherwise it is not equi-satisfiable. What remains is the question whether we need the unique bit-width to distinguish between different sorts. I am arguing that you can maintain a map of transformed terms to the original sort and maintain the mapping that way. I am quite certain that such a map is sufficient to identify the sort of the original term (i.e. there shouldn't be any collisions between different terms in that map) or at least I can't think of a case where it wouldn't be.

So for @yoni206's example:
```
(declare-sort S1 0)
(declare-sort S2 0)
(declare-const s11 S1)
(declare-const s12 S1)
(declare-const s21 S2)
(declare-const s22 S2)
(assert (= s11 s12))
(assert (not (= s21 s22)))
```

This would get turned into:
```
(declare-const bv11 (_ BitVec 1))
(declare-const bv12 (_ BitVec 1))
(declare-const bv21 (_ BitVec 1))
(declare-const bv22 (_ BitVec 1))
(assert (= bv11 bv12))
(assert (not (= bv21 bv22)))
```

And we would have a map:
```
bv11 --> S1
bv12 --> S1
bv21 --> S2
bv22 --> S2
```

Now let's say that we get the model:
```
bv11 = 0
bv12 = 0
bv21 = 0
bv22 = 1
```

We can use the map above to decide that `bv11` and `bv21` are representatives of different sorts, so we could get something like:
```
bv11 = r11
bv12 = r11
bv21 = r21
bv22 = r22
```

Can you think of an example where this wouldn't work?

Reply:
To give some additional intuition: We always replace terms of uninterpreted sorts with skolems (https://github.com/CVC4/CVC4/pull/3372/files#diff-4c3f5767ca5f0a1b1951e86d68f8afaeR221), skolems are unique w.r.t. the original sort that they belong to, so the map above should not have any collisions.

Reply:
I think you are right and this is possible. Intuitively, it is like using different "copies" of `BV[1]` for `S1` and `S2`, right? Without quantifiers, I also can't think of an example where this wouldn't work (but that doesn't mean much). @Ying1123 , I think this would complicate the correctness proof, but just a little bit.

Reply:
Yes, exactly. Instead of mapping the bit-width to the original sorts, we would be mapping the skolems themselves to the original sorts. I think the only additional piece that you need is to show that you can uniquely identify the original sort of each term in the formula and of course the formula needs to be well-typed (well-sorted? Not sure what's the correct terminology here).